File: Data Structures & Algorithms/jump-game/submission-7.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        # What is my greedy choice? -> whether I can reach my current goal from any previous index
        n=len(nums)
        farthest=0
        curr_end=0
        for i in range(n):
            farthest=max(farthest,nums[i]+i)
            if i == curr_end:
                curr_end=farthest
            
        return curr_end>=n-1
                


       
       
        
# O(N)

File: Data Structures & Algorithms/jump-game-ii/submission-7.go
func jump(nums []int) int {
    n:=len(nums)
	curr_end:=0
	jumps:=0
	farthest:=0
	for i:=0;i<n-1;i++ {
		if i+nums[i] > farthest {
			farthest = i+nums[i]
		}
		if i==curr_end{
			jumps+=1
			curr_end=farthest
		}
	}
	return jumps
}
